node/pkg/p2p/p2p.go

```diff
@@ -426,332 +497,618 @@ func Run(params *RunParams) func(ctx context.Context) error {
-			timer := time.NewTimer(time.Second * 2)
-			defer timer.Stop()
+							b, err := proto.Marshal(&msg)
+							if err != nil {
+								panic(err)
+							}
+							return b
+						}()
+
+						if GossipCutoverComplete() {
+							err = controlPubsubTopic.Publish(ctx, b)
+							p2pMessagesSent.WithLabelValues("control").Inc()
+							if err != nil {
+								logger.Warn("failed to publish heartbeat message", zap.Error(err))
+							}
+						} else if vaaPubsubTopic != nil {
+							err = vaaPubsubTopic.Publish(ctx, b)
+							p2pMessagesSent.WithLabelValues("old_control").Inc()
+							if err != nil {
+								logger.Warn("failed to publish heartbeat message to old topic", zap.Error(err))
+							}
+						}
 
+						p2pHeartbeatsSent.Inc()
+						ctr += 1
+					}
+				}
+			}()
+		}
+
+		// This routine processes messages received from the internal channels and publishes them to gossip. ///////////////////
+		// NOTE: The go specification says that it is safe to receive on a nil channel, it just blocks forever.
+		go func() {
 			for {
 				select {
 				case <-ctx.Done():
 					return
-				case <-timer.C:
-					timer.Reset(15 * time.Second)
-
-					// create a heartbeat
-					b := func() []byte {
-						DefaultRegistry.mu.Lock()
-						defer DefaultRegistry.mu.Unlock()
-						networks := make([]*gossipv1.Heartbeat_Network, 0, len(DefaultRegistry.networkStats))
-						for _, v := range DefaultRegistry.networkStats {
-							errCtr := DefaultRegistry.GetErrorCount(vaa.ChainID(v.Id))
-							v.ErrorCount = errCtr
-							networks = append(networks, v)
-						}
-
-						features := make([]string, 0)
-						if params.gov != nil {
-							features = append(features, "governor")
-						}
-						if params.acct != nil {
-							features = append(features, params.acct.FeatureString())
-						}
-						if params.ibcFeaturesFunc != nil {
-							ibcFlags := params.ibcFeaturesFunc()
-							if ibcFlags != "" {
-								features = append(features, ibcFlags)
+				case msg := <-params.gossipControlSendC:
+					if controlPubsubTopic != nil {
+						if GossipCutoverComplete() {
+							err := controlPubsubTopic.Publish(ctx, msg)
+							p2pMessagesSent.WithLabelValues("control").Inc()
+							if err != nil {
+								logger.Error("failed to publish message from control queue", zap.Error(err))
+							}
+						} else if vaaPubsubTopic != nil {
+							err := vaaPubsubTopic.Publish(ctx, msg)
+							p2pMessagesSent.WithLabelValues("old_control").Inc()
+							if err != nil {
+								logger.Error("failed to publish message from control queue to old topic", zap.Error(err))
 							}
 						}
-						if params.gatewayRelayerEnabled {
-							features = append(features, "gwrelayer")
-						}
-						if params.ccqEnabled {
-							features = append(features, "ccq")
+					} else {
+						logger.Error("received a message on the control queue when we do not have a control topic")
```

Comment:
If we are after the cutover time and we don't have a control channel, shouldn't we do something more drastic than log?

Reply:
That would actually be a program bug, so I changed it (and similar cases for the other topics) to a panic.

Reply:
that panic makes me nervous, but agree it would be a programming error. wonder if there's any safer way to set this up.

Reply:
After some offline discussion, it was agreed that the panic is reasonable since this function always creates the control topic if the control channel is not nil, so this would definitely be a program bug.